src/modelmanager.cpp

```diff
@@ -828,6 +839,52 @@ void ModelManager::watcher(std::future<void> exitSignal) {
     SPDLOG_LOGGER_INFO(modelmanager_logger, "Stopped model manager thread");
 }
 
+void ModelManager::cleanerRoutine(uint32_t resourcesCleanupIntervalSec, uint32_t sequenceCleanerIntervalMinutes, std::future<void> cleanerExitSignal) {
+    SPDLOG_LOGGER_INFO(modelmanager_logger, "Started cleaner thread");
+
+    uint32_t resourcesCleanupIntervalMiliseconds = resourcesCleanupIntervalSec * 1000;
+    uint32_t sequenceCleanerIntervalMiliseconds = sequenceCleanerIntervalMinutes * 60 * 1000;
+
+    FunctorResourcesCleaner functorResourcesCleaner{*this};
+    FunctorSequenceCleaner functorSequenceCleaner{globalSequencesViewer};
+
+    ovms::cleanerRoutine(resourcesCleanupIntervalMiliseconds, functorResourcesCleaner, sequenceCleanerIntervalMiliseconds, functorSequenceCleaner, cleanerExitSignal);
+
+    SPDLOG_LOGGER_INFO(modelmanager_logger, "Stopped cleaner thread");
+}
+
+void cleanerRoutine(uint32_t resourcesCleanupInterval, FunctorResourcesCleaner& functorResourcesCleaner, uint32_t sequenceCleanerInterval, FunctorSequenceCleaner& functorSequenceCleaner, std::future<void>& cleanerExitSignal) {
+    uint32_t currentResourcesWaitTime = resourcesCleanupInterval;
+    uint32_t currentSequenceWaitTime = sequenceCleanerInterval;
+    bool shouldCheckForSequenceCleanup = sequenceCleanerInterval != 0;
+    uint32_t currentWaitTime = (!shouldCheckForSequenceCleanup || currentResourcesWaitTime < currentSequenceWaitTime) ? currentResourcesWaitTime : currentSequenceWaitTime;
+
+    while (cleanerExitSignal.wait_for(std::chrono::milliseconds(currentWaitTime)) == std::future_status::timeout) {
+        SPDLOG_LOGGER_DEBUG(modelmanager_logger, "Cleanup check cycle begin");
+
+        currentResourcesWaitTime = (currentResourcesWaitTime - currentWaitTime) == 0 ? resourcesCleanupInterval : currentResourcesWaitTime - currentWaitTime;
+        if (shouldCheckForSequenceCleanup)
+            currentSequenceWaitTime = (currentSequenceWaitTime - currentWaitTime) == 0 ? sequenceCleanerInterval : currentSequenceWaitTime - currentWaitTime;
+        currentWaitTime = (!shouldCheckForSequenceCleanup || currentResourcesWaitTime < currentSequenceWaitTime) ? currentResourcesWaitTime : currentSequenceWaitTime;
+
+        if (currentResourcesWaitTime == resourcesCleanupInterval)
+            functorResourcesCleaner.cleanup();
+        if (currentSequenceWaitTime == sequenceCleanerInterval && shouldCheckForSequenceCleanup)
+            functorSequenceCleaner.cleanup();
+
+        SPDLOG_LOGGER_DEBUG(modelmanager_logger, "Cleanup check cycle end");
+    }
+}
+
+void ModelManager::cleanupResources() {
```

Comment:
Is there race condition here? Don't we need to lock resources before we remove? Other thread can insert it here (watcher thread)

Reply:
I suppose that is the case.